### DISCUSSION TITLE
Mobile applications and the use of `tryRotate`

### DISCUSSION THREAD

**Discussion by axxel** [zxing-cpp/zxing-cpp#725] [Polls]
In the context of the discussion about the proper default settings for `ReaderOptions` (#722) I did a little experimentation with the Android demo app on my Pixel 5. I noticed that the `tryRotate` setting is way more expensive on that device than I was aware of. It almost doubles the runtime per frame when every other option is 'full on'. That is not the case on my desktop at all. This made me discover an unnecessary performance penalty cause by the way rotated images are currently handled. Working on this could maybe lower the extra cost from 100% to 50%. But before spending time on this, I'd like to understand how the library is actually used in a mobile context in the real world.

If you have any experience or opinion regarding the subject "performance vs `tryRotate`", please leave a comment.

 * In the answers below "linear barcode" includes PDF417 in this case for internal and technical reasons.
 * 'horizontally' in the answers means that the bars are vertical but the code as a whole is horizontal (read left to right).

**Comment by ryantrem** [comment]
We try to support all barcode types in all orientations in either portrait or landscape (eg all the things).

**Comment by vkrause** [comment]
Same for us, user expectation is that all possible combinations work.

**Reply by axxel** [reply]
Thanks. So you never saw a reason to disable `tryRotate` for performance reasons? Same question to @ryantrem.

For me personally, I noticed that if I play round with the Android demo app, which I do now and then when I want to get some real-world usage experience in the grocery store or whatever, I instinctively hold the phone upright and in a "normal" orientation relative to the code. So I never really used the `tryRotate` flag and hence, did not notice its significant extra costs. It would be a pity if people had that on by default and payed 100% surplus in runtime/battery "for nothing" because they do the same thing as I do with my phone in hand.

I also noticed that the ZXing java library does not even have this feature.

**Reply by vkrause** [reply]
Not for the scanning from a video feed, no, especially not when in the hands of "normal" users that unlike you or me usually don't even see an orientation in barcodes I think. I can see that being potentially useful though if you have a more controlled setup or influence on the workflow in which the scanning happens.

**Reply by ryantrem** [reply]
Not outside of our perf tolerance (around 30fps on the camera feed for typical mobile devices).

**Reply by axxel** [reply]
Alright, could you both tell me what the image resolution is that you are using? I assume you are cropping your video feed then? The demo app with everything enabled, including multi-symbol-detection, takes 50ms per frame (20fps) with the full 1280x720 resolution on my Pixel 5. Without tryRotate, is is only 25ms.

**Reply by vkrause** [reply]
I don't actually know, this runs on random mobile phones in our case, and we don't have any restrictions or preprocessing there. When I measured this while implementing it some time ago on a several year old mid range phone the performance was perfectly fine, 30 fps or more. We don't have multi-symbol-detection enabled though, we only look for a single barcode at a time.

**Reply by axxel** [reply]
@vkrause so you don't know the resolution and don't know whether or not the input is cropped? Can you point me to the source of your mobile-video-feed-app-code? Also: do you look for *all* supported symbologies or maybe only the 2D ones?

**Reply by vkrause** [reply]
I don't know the resolution (that depends on the user device), I know we don't crop, sorry if that was unclear. The code is here: https://invent.kde.org/frameworks/prison/-/tree/master/src/scanner (in particular the video* files). We do have at least one app  looking for all symbologies, but also some that are e.g. QR or 2D only.

**Reply by axxel** [reply]
Alright. One interesting observation: you are still supporting zxing-cpp pre 1.4.0 but you dropped Qt5 support? ;)

**Reply by vkrause** [reply]
Qt5 support is in another branch, this is the not even released yet new major version code. But yes, we could probably also increase the ZXing minimum required version a bit by now.

**Reply by axxel** [reply]
I guess 2.0 would be a good base version, at least for your new major version. See also https://repology.org/project/zxing-cpp/versions

**Reply by ryantrem** [reply]
I checked things out in our code... first I forgot that we do our barcode scanning on a background thread. So we maintain a 60fps camera feed, and we only try to scan a barcode on a camera frame when the previous one has finished. We support different camera frame resolutions as well, depending on a user's needs. On my iPhone 8 (pretty old device), if I use resolution of 902x1440, and scan for all supported barcode types, then ZXing::ReadBarcodes takes about 268ms with try rotate on, and 184ms with it off (also various other decode hints are enabled). If it helps, we are also using zxing-cpp at commit a392ec6b, so a bit out of date.

**Reply by axxel** [reply]
@ryantrem Thanks for the performance measurements. A couple thoughts:
* the difference between tryRotate on and off is more in line with my expectations than what I saw on my Pixel phone
* the runtime of 268ms per frame is disappointingly long. I did a quick test with my old Pixel 3 which was released just one year after the iPhone 8 and it takes about 60ms per 1280x720 frame with all bells and whistles enabled. If I 'scale that' number by the difference in image resolution, that is still only 84ms not 268. I can't quite believe that the CPU in my Pixel 3 should indeed be 3x faster. Are you sure you have a properly optimized release build running?
* if you find the time, I'd be interested to learn if you see a performance change (in either direction) if you replace [this line](https://github.com/zxing-cpp/zxing-cpp/blob/f291f88214894f1851c5b82c789b05a0224d8cab/core/src/GlobalHistogramBinarizer.cpp#L113) with `#if 1`.
* if you have the means to do a quick performance profiling, I'd be very interested to learn where those 268ms are spent.

**Reply by ryantrem** [reply]
Apologies, here is some better info:
1. In an actual release build (🙂) on my iPhone 8 at 902x1440 resolution with try rotate on, I'm seeing about 26ms vs about 15ms with try rotate off.
2. I didn't see any difference in perf with the change you mentioned above.
3. Here is a screenshot of a profile of the interesting bits (this is 5 consecutive calls to `ReadBarcodes` just to smooth out any potential variance):
<img width="1453" alt="image" src="https://github.com/zxing-cpp/zxing-cpp/assets/5084643/923ec709-363a-42ed-8c6f-056b96e1397d">

**Reply by axxel** [reply]
> Apologies, here is some better info:
> 
>     1. In an actual release build (🙂) on my iPhone 8 at 902x1440 resolution with try rotate on, I'm seeing about 26ms vs about 15ms with try rotate off.

Those numbers make me envious of your iPhone hardware. ;) One thing to note when comparing those numbers: they depend heavily on the content of the image. A blurry white background will cause significantly less computation compared to a high contrast page of text.

>     2. I didn't see any difference in perf with the change you mentioned above.

If you feel like tinkering even more: what happens when you replace [this line](https://github.com/zxing-cpp/zxing-cpp/blob/de4149b57250bbe1b807ab5aeed7472a11deed7d/core/src/HybridBinarizer.cpp#L33) with `#if 0`?

>     3. Here is a screenshot of a profile of the interesting bits (this is 5 consecutive calls to `ReadBarcodes` just to smooth out any potential variance):

That looks about right to me and is yet another reminder how much there is potentially to be gained with a SIMD optimized run length encoder...

Question: do you input RGB data or gray scale data? If possible, setup your camera such that is produces e.g. `Yuv` pixels, even better when the three channels are in separate planes, such that you can feed the `Y` channel as a densely packed `Lum` image.
